### ISSUE TITLE
CUDA_ERROR_ILLEGAL_ADDRESS in toy training example

### ISSUE THREAD

**Issue by fachu000** [tensorflow/tensorflow#41891]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Inside virtual container:
uname -a
Linux 3558c7dc300b 4.15.0-74-generic #84-Ubuntu SMP Thu Dec 19 08:06:28 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): v2.2.0-rc4-8-g2b96f3662b 2.2.0
- Python version: Python 3.6.8
- CUDA/cuDNN version: 
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2019 NVIDIA Corporation
Built on Sun_Jul_28_19:07:16_PDT_2019
Cuda compilation tools, release 10.1, V10.1.243

- GPU model and memory:

NVIDIA DGX-2
16x NVIDIA Tesla V100 (32GB)
2x Intel Xeon Platinum 8168 2.7GHz 24C/48T
1.5TB RAM
30TB Internal NVME SSD

**Describe the current behavior**

Crashes. Same happens in [Colab](https://colab.research.google.com/drive/1akS92A3mbh_L-c978OsTHvTFAxnv22vV?usp=sharing) and in tf-nightly. 

See crash report in colab:

2020-07-30 08:06:02.330777: E tensorflow/stream_executor/cuda/cuda_event.cc:29] Error polling for event status: failed to query event: CUDA_ERROR_ILLEGAL_ADDRESS: an illegal memory access was encountered

**Describe the expected behavior**

It should work, as it does for other values of the batch and repeat parameters.

**Standalone code to reproduce the issue**

```python

# This is a simplified version of https://www.tensorflow.org/guide/distributed_training

m_training_data = tf.constant([[1.1,1.]], dtype=tf.double)
dataset = tf.data.Dataset.from_tensor_slices((m_training_data,)).repeat(8*1024).batch(1024)

mirrored_strategy = tf.distribute.MirroredStrategy()
dist_dataset = mirrored_strategy.experimental_distribute_dataset(dataset)

def train_step(input):
    A = tf.matmul(input[0], tf.linalg.matrix_transpose(input[0]))
    return tf.linalg.det(A)

@tf.function
def distributed_train_step(dist_inputs):
    loss = mirrored_strategy.run(train_step, args=(dist_inputs,))
    return mirrored_strategy.reduce(tf.distribute.ReduceOp.SUM, loss,
                         axis=None)

for dist_inputs in dist_dataset:
    print(distributed_train_step(dist_inputs))

```

**Other info / logs** 
[crash_1002.txt](https://github.com/tensorflow/tensorflow/files/5000057/crash_1002.txt)

**Comment by amahendrakar** [comment]
On running the Colab notebook linked, I was able to reproduce the issue with TF v2.2, TF v2.3 and TF-nightly. 

Code runs without any issues for `repeat(8*1024).batch(1024)`, but when changed to `repeat(256).batch(256)` session crashes. Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/5b23114e3bf0b6840fe601351026f527/41891-tf-nightly.ipynb#scrollTo=LbLfQOflzywD). Thanks!

**Comment by nikitamaia** [comment]
Changing labels as I don't think this is a dist-strat specific issue. The crash also occurs with Default Strategy.

**Comment by nikitamaia** [comment]
Hi @fachu000, can you clarify where you are seeing the crash? Is it for the `repeat(8*1024).batch(1024)` example or the `repeat(256).batch(256)` example? Like @amahendrakar I seem to get the crash only with `repeat(256).batch(256)`

I can also see the crash with the following reduced code
```
m_training_data = tf.constant([[1,1]], dtype=tf.double)
dataset = tf.data.Dataset.from_tensor_slices((m_training_data,)).repeat(256).batch(256)


for input in dataset:
    A = tf.matmul(input[0], tf.linalg.matrix_transpose(input[0]))
    det = tf.linalg.det(A)
```

**Comment by fachu000** [comment]
Hi @nikitamaia, 

Colab crashes with .repeat(256).batch(256)

In the server I use (system described above), it works with .repeat(256).batch(256) but crashes with .repeat(8*1024).batch(1024)

One difference may be that I use 4 GPUs whereas Colab uses 1. 

daniel

**Comment by nikitamaia** [comment]
Thanks for the clarification. I'm seeing different behavior based on where I run the code and whether I use MirroredStrategy or the Default strategy.
I am able to run `.repeat(256).batch(256)` on GCP with 2 GPUs with MirroredStrategy, but `.repeat(8*1024).batch(1024)` crashes. I see the reverse behavior with 1 GPU in Colab.

I also tried running on GPU with 1 GPU and both cases worked.
However, with 2 GPUs, and using the Default strategy I get a crash on the  `.repeat(256).batch(256)` but `.repeat(8*1024).batch(1024)`  works.

**Comment by sanjoy** [comment]
@rmlarsen here cuSolverDN is failing with `CUSOLVER_STATUS_EXECUTION_FAILED`:

```
...
W tensorflow/core/framework/op_kernel.cc:1773] OP_REQUIRES failed at determinant_op.cc:228 : Internal: tensorflow/core/util/cuda_solvers.cc:466: cuSolverDN call failed with status =6
W tensorflow/core/framework/op_kernel.cc:1773] OP_REQUIRES failed at determinant_op.cc:228 : Internal: tensorflow/core/util/cuda_solvers.cc:466: cuSolverDN call failed with status =6
E tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:725] failed to record completion event; therefore, failed to create inter-stream dependency
I tensorflow/stream_executor/stream.cc:4952] [stream=0x402c0130,impl=0x402be810] did not memcpy host-to-device; source: 0x7f5089649fc0
E tensorflow/stream_executor/stream.cc:338] Error recording event in stream: Error recording CUDA event: CUDA_ERROR_ILLEGAL_ADDRESS: an illegal memory access was encountered; not marking stream as bad, as the Event object may be at fault. Monitor for further errors.
E tensorflow/stream_executor/cuda/cuda_event.cc:29] Error polling for event status: failed to query event: CUDA_ERROR_ILLEGAL_ADDRESS: an illegal memory access was encountered
F tensorflow/core/common_runtime/gpu/gpu_event_mgr.cc:220] Unexpected Event status: 1
```

Any idea how to debug this further?  Is it a user error?  If so, probably TF should be returning a friendlier error message.

**Comment by sushreebarsa** [comment]
Was able to reproduce the issue with TF v2.5 & TF v2.6.0-dev20210606 in Colab ,please find the gist [here ](https://colab.research.google.com/gist/sushreebarsa/063363b65702fec5dc1e1e7e5c3ce8a7/untitled282.ipynb)..Thanks!

**Comment by sushreebarsa** [comment]
Closing this issue as it is fixed in latest version of TensorFlow. Please feel free to reopen the issue if you still have a concern. Thanks!

**Comment by tedwards-flf** [comment]
Is this fix in 2.5.0 or 2.6.0?

**Comment by animesh-wynk** [comment]
I am getting a similar issue in tensorflow 2.11.0
[here are more details](https://stackoverflow.com/questions/75813079/error-polling-for-event-status-failed-to-query-event-cuda-error-illegal-addres)
